File: day-1/clojure/day-1.clj
(ns aoc2019.day1)
(require '[clojure.string :as str])

;; PREPARE DATA
; Fetch data from the file.
(def s (slurp "./../day-1-input.txt"))

; Convert file string into a vector of integers.
(def vector-of-modules 
 (map 
  #(Integer/parseInt %) 
  (str/split-lines (str/trim s))))

;; FUNCTIONS
; Calculate fuel for a mass unit
(defn calc-fuel [mass]
  (- (quot mass 3) 2))

; Calculate fuel need for given amount of fuel
(defn fuel-for-fuel [total, initial-fuel] 
 (if (< initial-fuel 1)
   total
   (fuel-for-fuel (+ total initial-fuel) (calc-fuel initial-fuel) )))

; Calculate total amount of fuel for module
(defn fuel-for-module [module-mass] 
 (fuel-for-fuel 0, (calc-fuel module-mass)))

;; RESULT
(print (reduce + (map fuel-for-module vector-of-modules)))

File: day-1/javascript/day-1.js
// import { FileReader } from 'file-reader';
// import 'fuel_for_module' from './fuel_for_module'
const assert_equal = require('./../../test.js')
const fs = require('fs');



const calc_fuel = (mass) => {
  let fuel = Math.floor(parseInt(mass) / 3) - 2

  return fuel < 0 ? 0 : fuel;
};

let fuel_for_fuel = function (total, initial_fuel) {
  if (initial_fuel < 1) {
    return total;
  } else {
    return fuel_for_fuel(total + initial_fuel, calc_fuel(initial_fuel));
  }
};

const fuel_for_module = module_mass => {
  return fuel_for_fuel(0, calc_fuel(module_mass));
}

fs.readFile('./../day-1-input.txt', 'utf8', function(err, contents) {
  
  const array_of_modules = contents.split(/\n/)

  const result = array_of_modules.map(fuel_for_module).reduce((acc, curr) => acc + curr)

  console.log(result);
});

// TEST
assert_equal(calc_fuel(23), 5, 'calc_fuel test');
assert_equal(fuel_for_fuel(0, 100), 139, 'fuel_for_fuel test')
assert_equal(fuel_for_module(36), 11, 'fuel_for_module test')
